Simplify menu code, fix PageDown, move up only on Up, 'k' and BackTab

proski opened the pull request:
The new code is much closer to the upstream and has been verified again the upstream code. Some discrepancies have been fixed. Most importantly, keys don't move the cursor up unless they are specifically meant to do it.

Two issues have been found upstream. The fixes are in this PR. Upstream PRs:
https://github.com/tmux/tmux/pull/4580
https://github.com/tmux/tmux/pull/4581

proski commented:
To make CodeQL happy, I've added another commit that removes the function-level `name` and replaces it with narrow-scoped `name` variables. In one case (NPAGE), I had to add an assignment for `name` that is never used. It suppresses a Rust compiler warning. And it's exactly where CodeQL was unhappy. Apparently, they could not figure out that if `i` starts with 5, `while i > 0` would run at least once and initialize `name`.

I would just rewrite that PageUp/PageDown/Home/End code to avoid backtracking after the move and to avoid micro-optimization that adds complexity, but we want to stay in sync with tmux for now.

proski commented:
Replaced the workaround with a cleaner approach. `while i > 0` is now `loop`. That's what we do in many other places.

richardscollin commented:
I haven't had a chance to look at this yet, but don't worry about codeql. I care more about if the code is actually correct / the quality of the code.

There are many existing codeql issues, but I think they are private by default, most seem to be false positives. I enabled codeql because I was curious about the tool and maybe trying to integrate my own lints so they have a similar ui in pull requests. I don't think it's really useful for this project yet.

proski commented:
Good to know. I still think it was a good change to remove the function-level `name`.

proski commented:
I've updated this PR to minimize differences from the corresponding tmux code (master branch). Most notably, I added another labeled scope, `'tab_down_j`, to make it possible to use a conditional `break` that breaks from the `switch` in C and `match` in Rust.

Comparing to the original Rust code is also possible by selecting Gear->Hide whitespace.

proski commented:
Better yet, I wrapped the whole `match` in a block labeled as `'match_`. It turns out the upstream has another conditional break, so it's good to have a way to break from any match arm. Sure, it increases the indent, but the goal for now is to have an accurate representation of the upstream code.
